src/box/lua/load_cfg.lua

```diff
@@ -187,6 +188,57 @@ local default_cfg = {
     txn_isolation         = "best-effort",
 }
 
+-- We need to track cfg changes done through API of distinct modules (log.cfg of
+-- log module for example). We cannot use just box.cfg because it is not
+-- available before box.cfg() call and other modules can be configured before
+-- this moment.
+local pre_load_cfg = table.copy(default_cfg)
+
+-- On first box.cfg{} we need to know options that were already configured
+-- in standalone modules (like log module). We should not apply env vars
+-- for these options. pre_load_cfg is not suitable for this purpose because
+-- of nil values.
+local pre_load_cfg_is_set = {}
+
+local replication_sync_timeout_brief = [[
+Sets the default value for box.cfg.replication_sync_timeout.
+Old is 300 seconds, new is 0 seconds. New behaviour makes
+box.cfg{replication = ...} call exit without waiting for
+synchronisation with all the remote nodes. This means that the node
+might be in 'orphan' state for some time after the box.cfg{} call
+returns. Set before first box.cfg{} call in order for the option to take effect.
+
+https://github.com/tarantool/tarantool/wiki/compat%3Abox_cfg_replication_sync_timeout
+]]
+
+-- A list of box.cfg options whose defaults are managed by compat.
+local compat_options = {
+    {
+        name = 'replication_sync_timeout',
+        brief = replication_sync_timeout_brief,
+        oldval = 300,
+        newval = 0,
+        obsolete = nil,
+        default = 'old',
+    },
+}
+
+for _, option in ipairs(compat_options) do
+    local option_name = 'box_cfg_' .. option.name
+    compat.add_option({
+        name = option_name,
+        default = option.default,
+        obsolete = option.obsolete,
+        brief = option.brief,
+        action = function(is_new)
+            local val = is_new and option.newval or option.oldval
+            default_cfg[option.name] = val
+            pre_load_cfg[option.name] = val
+        end,
```

Comment:
We can raise an error on attempt to set the option after/during first `box.cfg()`. It should work better than a caution in the help message :)

It seems, it should be something like this:

```lua
local function is_locked()
    return lock:is_empty()
end

if box_is_configured or is_locked() then
    error(<...>)
end
```

Reply:
Good idea, done:
<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/load_cfg.lua b/src/box/lua/load_cfg.lua
index 0c5bca6f5..4330ca00d 100644
--- a/src/box/lua/load_cfg.lua
+++ b/src/box/lua/load_cfg.lua
@@ -25,6 +25,10 @@ local function locked(f)
     end
 end
 
+local function is_locked()
+    return lock:is_full()
+end
+
 --
 -- When a feature is disabled, every single mentioning of it should
 -- be eliminated. Even box.cfg{} should not accept any parameters
@@ -200,6 +204,17 @@ local pre_load_cfg = table.copy(default_cfg)
 -- of nil values.
 local pre_load_cfg_is_set = {}
 
+
+-- Whether box is loaded.
+--
+-- `false` when box is not configured or when the initialization
+-- is in progress.
+--
+-- `true` when box is configured.
+--
+-- Use locked() wrapper to obtain reliable results.
+local box_is_configured = false
+
 local replication_sync_timeout_brief = [[
 Sets the default value for box.cfg.replication_sync_timeout.
 Old is 300 seconds, new is 0 seconds. New behaviour makes
@@ -231,6 +246,10 @@ for _, option in ipairs(compat_options) do
         obsolete = option.obsolete,
         brief = option.brief,
         action = function(is_new)
+            if is_locked() or box_is_configured then
+                error("The compat  option '" .. option_name .. "' takes " ..
+                      "effect only before the initial box.cfg() call")
+            end
             local val = is_new and option.newval or option.oldval
             default_cfg[option.name] = val
             pre_load_cfg[option.name] = val
@@ -952,16 +971,6 @@ setmetatable(box, {
      end
 })
 
--- Whether box is loaded.
---
--- `false` when box is not configured or when the initialization
--- is in progress.
---
--- `true` when box is configured.
---
--- Use locked() wrapper to obtain reliable results.
-local box_is_configured = false
-
 local function load_cfg(cfg)
     -- A user may save box.cfg (this function) before box loading
     -- and call it afterwards. We should reconfigure box in the
```
</details>